### ISSUE TITLE
components should not be exposed via window

### ISSUE THREAD

**Issue by ohadlevy** [ManageIQ/react-ui-components#2]
imho, we should not continue using window as a way to share code, but rather use webpack import to handle it for us. (https://github.com/ManageIQ/react-ui-components/blob/master/src/vendor.js#L12)

the main motivation for me is to find errors during build time vs run time.

**Comment by karelhala** [comment]
Whenever we see component accessing window object we can't merge such PR, however in vendor.js we store some libraries into window object so we can access them via import like this `import * as React from 'react'` and when we use [externals/#object](https://webpack.js.org/configuration/externals/#object) final build will not include these libraries.

I couldn't find better sollution to access these libraries exposed with externals and they ended up as part of our build, which is not acceptable.

**Comment by ohadlevy** [comment]
I'm not sure i follow, which error did you face? you can always use webpack to build and reference via alias the path to find those libraries.

btw, having said that, since this repo should be consumed as a package, imho it should not import its own redux/react etc as it would most likely conflict with the packages the core repo requires.

**Comment by karelhala** [comment]
> it should not import its own redux/react etc as it would most likely conflict with the packages the core repo requires

That's exactly the issue I was facing. We have to import react/redux/lodash/whatever so we can check errors during build time, but we can't include them into final build but import these packages from application which imports this library. That's what https://webpack.js.org/configuration/externals/ is for. However some libraries were not behaving correctly and I could not access them, webpack tried to outsmart me and included them as part of build (which wasn't correct), so I stored these packages into window object and configured webpack in a way that he will load these packages from window.

So long story short
 * we can't import other libraries into final build - use externals.
 * some libraries can't be loaded correctly using externals - store these libraries into window object.

**Comment by ohadlevy** [comment]
TBH I disagree, that's why you have peer dependencies in your package.json file.

**Comment by karelhala** [comment]
You are right, that we should add `react` and other libraries to peerDependencies, but still when running build with webpack these dependendencies would be bunled with our library, which is no go. We also don't want to watch for these external libraries because build would be too slow.

So the problem is not to include external libraries into final build, but we want to use imports and not access window properties.

I tried couple of approaches
 * Using https://webpack.js.org/plugins/commons-chunk-plugin/ - won't work because in final build webpack would try to load them, but they'll be missing.
 * Using separate build process for vendor's libraries - using ES6's `import` is not working.
 * Loading external libraries from CDN - using ES6's `import` is not working.

Best solution to this looks like using externals is the best way, we however have to use object defintion to our externals. So we can cover all load mechanisms, notice how webpack is then loading these libraries https://github.com/webpack/webpack/tree/master/examples/externals#distoutputjs

But again, you are correct we can't access window's libraries (from components), eslint helps us finding if someone is not using `import`, those vendor libraries however have to be accessible trough window, so webpack can work with them.

**Comment by himdel** [comment]
This is probably a moot point anyway, we do need React on window in ui-classic anyway.

So, if the library expects it there, so be it :).

I'm not proposing we do this for ever single dependency, but for the major ones, there is really no benefit in having `import * as _ from 'lodash'` and `import * as React from 'react`' etc, etc. in every single file.

**Comment by ohadlevy** [comment]
can you explain why ui-classic requires it? not sure that i follow.

AFAIU, manageiq webpack should build the end result, in which, there would
be only one copy of the library. I'm not sure if this repo should even have
an official build (besides the need for a storybook).

**Comment by karelhala** [comment]
I think we diverged a lot from initial issue. Let's summarize things which let me to include external libraries into window. 
* We have to expose this as library - use webpack's settings to output library https://webpack.js.org/configuration/output/#output-library
* We need to watch for errors during build - we need to use ES6's imports, or require
* We can't include external 3rd party libraries into the build - use webpack's external (https://webpack.js.org/configuration/externals/ it's well described here why to use externals) and a lot of examples and others recommend this settings for libraries
* Some libraries were not behaving correctly with external - when loading external library we can specify 4 different ways of how to load them (this shows how webpack is awesome), the 4th option specifies loading from root, which is window.

@ohadlevy as you mentioned, I tried using webpack imports, it was actually what we done in https://github.com/manageIq/ui-components however there was angular in play which helped us with loading of these modules. Here on other side we expose this as library, which is a bit harder to maintain and we have to use externals.

> I'm not sure if this repo should even have an official build

Well technically we don't **need** to build it and let ui-classic to bundle it together, but that doesn't feel right. We'd have to change build process in ui-classic and rely on it, which is not good. This repo would eventually be just code which could be part of ui-classic. But what if in future someone would like to use any component from this repo? Simplest way would be to copy it over.

This settings is actually really great. For example simple component which uses patternfly-react alert with some label on top of it has 6 KBs of unminified code. When you want to use this component in ui-classic (but it can be any application as long as it uses commonjs, commonjs2, AMD or stores external libraries in window object) you'll do something like this:
```javascript
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { Alert } from 'patternfly-react';
import { AlertExample } from '@manageiq/react-ui-components/dist/comp-a';

document.addEventListener("DOMContentLoaded", function(event) { 
  ReactDOM.render(AlertExample, document.getElementById('alert-to-show'));
});
```
All things just click together, webpack does all the work, since npm installs our dependencies (TODO - list core dependencies in depenencies) and we already set up externals in a way webpack can find them it just works. Believe me I tried a lot of approaches and this one feels best and is great. We could use some magic loader like https://github.com/webpack-contrib/imports-loader, but then we'd have to list all of these dependencies in webpack config, which is confusing.

**Comment by ohadlevy** [comment]
@karelhala i'm sorry that I'm not getting into the details... but how is this build is any different to the patternfly-react build? they don't force me to use window and import just works...?

**Comment by karelhala** [comment]
They don't use any bundler, they use plain babel, it's fine for them, but it would suit us that great. The thing is that patternfly-react can't use dead code detection, minification and other things which come as part of webpack. Another thing is that they have to use specific loading of sass files, again, nothing wrong on this, but webpack helps us with these imports. They need to have multiple script tasks just to lint js, scss and less files. It's fine for them as far as I know they hed discussion about this and they went with plain simple npm tasks for these builds. Those tasks are fine, but I prefer using webpack because of his goodies.

And this settings doesn't force you in any way to use winow, it's just to cover all possible ways of how to load certain libraries. Plus the vendor is just for demo application and for storybook so we have one place where we store these external libraries.

**Comment by himdel** [comment]
I think we can close this one.

We're not exposing actual components on window,
and in fact, the only code here setting stuff window is in `src/global-functions/test/translationMock.test.js`, and `src/vendor.js`, neither of which get loaded by the ui.
